packages/amesos2/src/Amesos2_TpetraMultiVecAdapter_def.hpp

```diff
@@ -219,6 +219,133 @@ namespace Amesos2 {
+    if ( num_vecs == 1 && this->getComm()->getRank() == 0 && this->getComm()->getSize() == 1 ) {
+      // MDM-TODO Check and discuss - need to learn more about MV
+      // WARNING - This is not tested currently by Tacho Amesos2_Solver_Test - tested manually by setting numVecs 1 in Solver_Test.cpp. Add param?
+      kokkos_view = mv_->template getLocalView<typename KV::execution_space>();
+    }
+    else {
+
+      // (Re)compute the Export object if necessary.  If not, then we
+      // don't need to clone distribution_map; we can instead just get
+      // the previously cloned target Map from the Export object.
+      RCP<const map_type> distMap;
+      if (exporter_.is_null () ||
+          ! exporter_->getSourceMap ()->isSameAs (* (this->getMap ())) ||
+          ! exporter_->getTargetMap ()->isSameAs (* distribution_map)) {
+        // Since we're caching the Export object, and since the Export
+        // needs to keep the distribution Map, we have to make a copy of
+        // the latter in order to ensure that it will stick around past
+        // the scope of this function call.  (Ptr is not reference
+        // counted.)
+        distMap = rcp(new map_type(*distribution_map));
+        // (Re)create the Export object.
+        exporter_ = rcp (new export_type (this->getMap (), distMap));
+      }
+      else {
+        distMap = exporter_->getTargetMap ();
+      }
+
+      multivec_t redist_mv (distMap, num_vecs);
+
+      // Redistribute the input (multi)vector.
+      redist_mv.doExport (*mv_, *exporter_, Tpetra::REPLACE);
+
+      if ( distribution != CONTIGUOUS_AND_ROOTED ) {
+        // Do this if GIDs contiguous - existing functionality
+
+        // MDM-TODO Check and discuss - need to learn more about MV
+        kokkos_view = redist_mv.template getLocalView<typename KV::execution_space>();
+      }
+      else {
+        // Do this if GIDs not contiguous...
+        // sync is needed for example if mv was updated on device, but will be passed through Amesos2 to solver running on host
+        typedef typename multivec_t::dual_view_type dual_view_type;
+        typedef typename dual_view_type::host_mirror_space host_execution_space;
+        redist_mv.template sync < host_execution_space > ();
+
+        auto contig_local_view_2d = redist_mv.template getLocalView<host_execution_space>();
+        if ( redist_mv.isConstantStride() ) {
+          // MDM-TODO - Determine if this is the best way
+          // WARNING - This is not tested currently by Tacho Amesos2_Solver_Test so not tested anywhere yet!
+          auto contig_local_view_2d = redist_mv.template getLocalView<typename KV::execution_space>();
+          Kokkos::parallel_for(
+            Kokkos::RangePolicy<typename KV::execution_space, size_t> (0, num_vecs * lda),
+            KOKKOS_LAMBDA (const size_t & n) {
+            auto i = n % num_vecs;
+            auto j = n / num_vecs;
+            kokkos_view(i,j) = contig_local_view_2d(i,j);
```

Comment:
You don't need to write a kernel here; just use `Kokkos::deep_copy`.

Reply:
Changed to deep_copy.